File: mm-kt/src/main/kotlin/io/github/metamessage/mm/MimeWire.kt
package io.github.metamessage.mm

object MimeWire {
    private val STR_TO_MIME = mutableMapOf<String, Int>().apply {
        put("image/jpeg", 1)
        put("image/jpg", 1)
        put("image/png", 2)
        put("image/gif", 3)
        put("image/webp", 4)
        put("image/svg+xml", 5)
        put("image/avif", 6)
        put("image/bmp", 7)
        put("image/x-icon", 8)
        put("image/tiff", 9)
        put("image/heic", 10)
        put("image/heif", 11)
        put("text/plain", 12)
        put("text/html", 13)
        put("text/css", 14)
        put("text/javascript", 15)
        put("application/javascript", 15)
        put("application/json", 16)
        put("text/csv", 17)
        put("text/markdown", 18)
        put("application/pdf", 19)
        put("application/zip", 20)
        put("application/gzip", 21)
        put("application/x-tar", 22)
        put("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet", 23)
        put("application/vnd.openxmlformats-officedocument.wordprocessingml.document", 24)
        put("application/octet-stream", 25)
        put("video/mp4", 26)
        put("video/webm", 27)
        put("video/mov", 28)
        put("audio/mpeg", 29)
        put("audio/wav", 30)
        put("audio/flac", 31)
        put("font/woff2", 32)
        put("font/ttf", 33)
    }

    fun parse(s: String?): Int {
        if (s.isNullOrBlank()) return 0
        return STR_TO_MIME[s.trim().lowercase()] ?: 0
    }

    fun toString(code: Int): String = when (code) {
        1 -> "image/jpeg"
        2 -> "image/png"
        3 -> "image/gif"
        4 -> "image/webp"
        5 -> "image/svg+xml"
        6 -> "image/avif"
        7 -> "image/bmp"
        8 -> "image/x-icon"
        9 -> "image/tiff"
        10 -> "image/heic"
        11 -> "image/heif"
        12 -> "text/plain"
        13 -> "text/html"
        14 -> "text/css"
        15 -> "text/javascript"
        16 -> "application/json"
        17 -> "text/csv"
        18 -> "text/markdown"
        19 -> "application/pdf"
        20 -> "application/zip"
        21 -> "application/gzip"
        22 -> "application/x-tar"
        23 -> "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
        24 -> "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
        25 -> "application/octet-stream"
        26 -> "video/mp4"
        27 -> "video/webm"
        28 -> "video/mov"
        29 -> "audio/mpeg"
        30 -> "audio/wav"
        31 -> "audio/flac"
        32 -> "font/woff2"
        33 -> "font/ttf"
        else -> "unknown"
    }
}

File: mm-cs/src/MetaMessage/Core/MimeWire.cs
namespace MetaMessage.Core;

public static class MimeWire
{
    private static readonly Dictionary<string, int> MimeToIndex = new(StringComparer.OrdinalIgnoreCase)
    {
        {"image/jpeg", 1},
        {"image/jpg", 1},
        {"image/png", 2},
        {"image/gif", 3},
        {"image/webp", 4},
        {"image/svg+xml", 5},
        {"image/avif", 6},
        {"image/bmp", 7},
        {"image/x-icon", 8},
        {"image/tiff", 9},
        {"image/heic", 10},
        {"image/heif", 11},
        {"text/plain", 12},
        {"text/html", 13},
        {"text/css", 14},
        {"text/javascript", 15},
        {"application/json", 16},
        {"text/csv", 17},
        {"text/markdown", 18},
        {"application/pdf", 19},
        {"application/zip", 20},
        {"application/gzip", 21},
        {"application/x-tar", 22},
        {"application/vnd.openxmlformats-officedocument.spreadsheetml.sheet", 23},
        {"application/vnd.openxmlformats-officedocument.wordprocessingml.document", 24},
        {"application/octet-stream", 25},
        {"video/mp4", 26},
        {"video/webm", 27},
        {"video/mov", 28},
        {"audio/mpeg", 29},
        {"audio/wav", 30},
        {"audio/flac", 31},
        {"font/woff2", 32},
        {"font/ttf", 33},
    };

    public static int ParseMIME(string s)
    {
        s = s.Trim().ToLowerInvariant();
        if (MimeToIndex.TryGetValue(s, out int index))
        {
            return index;
        }
        return 0;
    }
    public static string MimeForWire(string mimeType)
    {
        return mimeType switch
        {
            "application/json" => "json",
            "application/xml" => "xml",
            "text/plain" => "txt",
            "text/html" => "html",
            "text/css" => "css",
            "text/javascript" => "js",
            "image/jpeg" => "jpg",
            "image/png" => "png",
            "image/gif" => "gif",
            "image/webp" => "webp",
            "audio/mpeg" => "mp3",
            "audio/wav" => "wav",
            "video/mp4" => "mp4",
            "video/webm" => "webm",
            "application/pdf" => "pdf",
            "application/zip" => "zip",
            "application/gzip" => "gz",
            "application/tar" => "tar",
            "application/7z" => "7z",
            "text/csv" => "csv",
            "application/vnd.ms-excel" => "xls",
            "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet" => "xlsx",
            "application/msword" => "doc",
            "application/vnd.openxmlformats-officedocument.wordprocessingml.document" => "docx",
            "application/vnd.ms-powerpoint" => "ppt",
            "application/vnd.openxmlformats-officedocument.presentationml.presentation" => "pptx",
            _ => mimeType
        };
    }

    public static string WireForMime(string wireMime)
    {
        return wireMime switch
        {
            "json" => "application/json",
            "xml" => "application/xml",
            "txt" => "text/plain",
            "html" => "text/html",
            "css" => "text/css",
            "js" => "text/javascript",
            "jpg" => "image/jpeg",
            "png" => "image/png",
            "gif" => "image/gif",
            "webp" => "image/webp",
            "mp3" => "audio/mpeg",
            "wav" => "audio/wav",
            "mp4" => "video/mp4",
            "webm" => "video/webm",
            "pdf" => "application/pdf",
            "zip" => "application/zip",
            "gz" => "application/gzip",
            "tar" => "application/tar",
            "7z" => "application/7z",
            "csv" => "text/csv",
            "xls" => "application/vnd.ms-excel",
            "xlsx" => "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            "doc" => "application/msword",
            "docx" => "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
            "ppt" => "application/vnd.ms-powerpoint",
            "pptx" => "application/vnd.openxmlformats-officedocument.presentationml.presentation",
            _ => wireMime
        };
    }
}